pkg/proxy/endpointslicecache.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"sort"
+
+	discovery "k8s.io/api/discovery/v1alpha1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+	utilproxy "k8s.io/kubernetes/pkg/proxy/util"
+	utilnet "k8s.io/utils/net"
+)
+
+// EndpointSliceCache is used as a cache of EndpointSlice information
+type EndpointSliceCache struct {
+	slicesByService  map[types.NamespacedName]map[string]EndpointSliceInfo
+	makeEndpointInfo makeEndpointFunc
+	hostname         string
+	isIPv6Mode       *bool
+	recorder         record.EventRecorder
+}
+
+// EndpointSliceInfo contains just the attributes kube-proxy cares about.
+// Used for caching. Intentionally small to limit memory util.
+type EndpointSliceInfo struct {
+	Ports     []discovery.EndpointPort
+	Endpoints []EndpointInfo
+}
+
+// EndpointInfo contains just the attributes kube-proxy cares about.
+// Used for caching. Intentionally small to limit memory util.
+type EndpointInfo struct {
+	Addresses []string
+	Topology  map[string]string
+}
+
+// NewEndpointSliceCache initializes an EndpointSliceCache
+func NewEndpointSliceCache(hostname string, isIPv6Mode *bool, recorder record.EventRecorder, makeEndpointInfo makeEndpointFunc) *EndpointSliceCache {
```

Comment:
There are two feature gates now. The one we already had is at a very high level in the kube-proxy startup where we initialize the endpoint slice informer and handlers if the feature gate is enabled, otherwise we enable the traditional endpoints logic. https://github.com/robscott/kubernetes/blob/endpointslice-proxy/cmd/kube-proxy/app/server.go#L621

I added a second feature gate check in the ipvs and iptables proxiers that gets passed through to NewEndpointChangeTracker() and only initializes an EndpointSliceCache if that feature gate is set. Hopefully those two are sufficient.